From 459663ac7d7d8d2c44d9af726656c9689e660798 Mon Sep 17 00:00:00 2001
From: Yannick Fertre <yannick.fertre@foss.st.com>
Date: Mon, 19 Feb 2024 15:16:34 +0100
Subject: [PATCH 0937/1141] drm/stm: check simple-framebuffer state

To avoid conflicts between the simple-framebuffer and the
display-controller, a check was added concerning the state
of the simple-framebuffer (must be probed).

Change-Id: I4fbafa7cd53436ef77ecdcdd0ebf06f1e83b33ec
Signed-off-by: Yannick Fertre <yannick.fertre@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/361006
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
---
 drivers/gpu/drm/stm/drv.c | 23 ++++++++++++++++++++++-
 1 file changed, 22 insertions(+), 1 deletion(-)

--- a/drivers/gpu/drm/stm/drv.c
+++ b/drivers/gpu/drm/stm/drv.c
@@ -177,10 +177,31 @@ static int stm_drm_platform_probe(struct
 	struct device *dev = &pdev->dev;
 	struct drm_device *ddev;
 	struct ltdc_device *ldev;
-	int ret;
+	struct device *sfdev;
+	struct device_node *node;
+	int ret = 0;
 
 	DRM_DEBUG_DRIVER("\n");
 
+	/*
+	 * To avoid conflicts between the simple-framebuffer and the display-controller,
+	 * a check was added concerning the state of the simple-framebuffer (must be probed).
+	 */
+	node = of_find_compatible_node(NULL, NULL, "simple-framebuffer");
+	if (!IS_ERR(node)) {
+		if (of_device_is_available(node)) {
+			sfdev = bus_find_device_by_of_node(&platform_bus_type, node);
+			if (sfdev) {
+				if (!device_is_bound(sfdev))
+					ret = -EPROBE_DEFER;
+				put_device(sfdev);
+			}
+		}
+		of_node_put(node);
+		if (ret)
+			return ret;
+	}
+
 	ldev = devm_kzalloc(dev, sizeof(*ldev), GFP_KERNEL);
 	if (!ldev)
 		return -ENOMEM;
